### DISCUSSION TITLE
Use mDNS to support zero-configuration for apollo client?

### DISCUSSION THREAD

**Discussion by nobodyiam** [apolloconfig/apollo#4338] [Ideas]
Apollo client relies on the meta server to locate the config service addresses, which requires the user to [configure](https://www.apolloconfig.com/#/en/usage/java-sdk-user-guide?id=_122-apollo-meta-server) the meta server address properly.

![image](https://user-images.githubusercontent.com/837658/167048394-c4d8edb2-7b71-45af-bffb-ea16859cf600.png)

It's not a very complex configuration, but it does increase some burden and it's easy to make mistakes as different environments have different meta server addresses.
However, with the concept of [Multicast DNS(mDNS)](https://en.wikipedia.org/wiki/Multicast_DNS), it's possible to achieve the zero-configuration goal: the apollo client simply uses `http://apollo.local` to access apollo meta server and it requires no configuration, e.g. no need to configure the DNS server for this domain and users don't need to configure the meta server address when using apollo-client.

![image](https://user-images.githubusercontent.com/837658/167050101-89dda563-6afa-4439-ae3e-a8dc27254346.png)

**Comment by Anilople** [comment]
Sounds great!

**Comment by nisiyong** [comment]
If there are several config-services, I think the client could not recognize the right environment one.

**Reply by nobodyiam** [reply]
This is a potential issue. Each environment should have a distinct LAN, or the mDNS request would be sent to apollo-configservices in multiple environments.

**Comment by qmwu2000** [comment]
It works well for strictly managed environments, such as production or QA environment, in a low priority. For development environment in the office or even in local dev box, Apollo root configure address should/could be overridden by existing methods.

**Reply by nobodyiam** [reply]
Right, we could keep using the current meta address locate logic and fallback to mDNS if no address could be found.
